Usability improvements for Kafka Streams

gunnarmorling opened the pull request:
* Taking care of lifecycle management in extension
* Allowing to configure Kafka Streams through application.properties

Hey @cescoffier, @gsmet et al., here's the first follow-up PR for the Kafka Streams extension. It makes KS much simpler to use by taking care of the lifecycle automatically from within the extension and also by making things configurable through _application.properties_. This will also allow to automatically expose healthchecks and metrics in the next step. Feedback welcome!

gunnarmorling committed:
#2863 Usability improvements for Kafka Streams

* Taking care of lifecycle management in extension
* Allowing to configure Kafka Streams through application.properties

gunnarmorling commented on integration-tests/main/src/main/resources/application.properties:
```diff
@@ -18,10 +18,6 @@ quarkus.resteasy.gzip.max-input=10
 web-message=A message
 schedulerservice.cron.expr=0/10 * * * * ?
 
-quarkus.log.category.kafka.level=WARN
```
This was a left-over from the earlier extraction of the tests into their own module.

gunnarmorling commented on integration-tests/kafka/src/main/resources/application.properties:
```diff
@@ -1,3 +1,13 @@
+quarkus.kafka.streams.topics=streams-test-categories,streams-test-customers
+
+# streams options
+quarkus.kafka.streams.application.id=streams-test-pipeline
```
I'm getting warnings due to these options being "unknown" in the Quarkus namespace. I'm not quite sure how to handle this, in particular as it's not a fixed set of known options, but it could be any Kafka client/producer/streams option really that'll just be passed through. I thought though that it'd be the "Quarkus way" to have this under some sort of common prefix". Any guidance would be appreciated!

machi1990 replied:
Hi @gunnarmorling, here are my2cents about this:

I think you'll need a RUNTIME config 
```
@ConfigRoot(name="kafka", phase=RUN_TIME)
public class KafkaStreamConfig {
    // streams which is a config 
   public KafkaStreams streams;
}

@ConfigGroup
pubic KafkaStreams {
 .. config items here
}

```
See `agroal` extension on how to define them: https://github.com/quarkusio/quarkus/blob/a91a034b855cae3f8d474951cb33e8d05707dc13/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalRuntimeConfig.java#L9. 

Then you can pass around you configs a templates (recorder)  till `KafkaStreamsTopologyManager ` I guess that's where they are used. 

Also you can check hibernate extension which defines several levels of config: https://github.com/quarkusio/quarkus/blob/3e14f37748288c44d531b57d1c80fb7c887b0c02/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfig.java#L13

geoand replied:
I'm not sure what the proper solution is, but it is possible (although a crappy solution) to remove this false positives: https://github.com/quarkusio/quarkus/blob/00e2bff5b08d9a99743d0104bfd692602bcdeb0e/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java#L65

geoand replied:
But looking at the test failure, it looks like something is entirely wrong with the config. I think @dmlloyd needs to give the expert solution here

gunnarmorling replied:
Yes, that's what I'm wondering, too. I'd need to know why the values aren't exposed in native mode. On the "crappy solution" you linked to, the problem is that the set of these options is dynamic, i.e. it could really be any option known to Kafka consumers/producers/streams, and it's not realistic to hard-wire all of them in our code.

geoand replied:
Sure, but we could just add wildcard support or something to the crappy solution, although that would just make it even crappier probably...

gsmet replied:
Ah, so you already had the discussion here, nice :).

So... about the options, maybe we could at least try to have the most common ones hard wired. It would be more consistent. The issue is that it probably wouldn't be consistent with the more advanced ones. Not sure the notion of "most common ones" make sense here?

For the wildcard solution, you need to use a `Map` in the config object. I think it would work OK but I never used it to point to values, not config objects. So we would have to test it.

What I don't like with all this is that you end up with a bunch of `quarkus` totally inconsistent with our usual naming (with `dashes`) and our convention for types (Duration comes to mind).

Not sure there's an ideal solution there, though.

gunnarmorling replied:
> So... about the options, maybe we could at least try to have the most common ones hard wired. It would be more consistent.

@gsmet, WDYM by hard-wired here?

> you need to use a Map in the config object.

Do you have an example for such config object?

> What I don't like with all this...

I'm open to alternatives. I moved this under the "io.quarkus" namespace because it felt like the way to do it, but we also could just support giving the options with their original names in application.properties. Is there precedent for that?

I'm not sure about the dashes and types part; keep in mind all these are options known to and used by Kafka Streams users. If we alter names or value representations, it'll create friction when migrating existing code or simply applying the name/value someone has read in the upstream docs or other material.

gsmet commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -0,0 +1,167 @@
+
+    private static final String BOOTSTRAP_SERVERS_PROPERTY = "kafka.bootstrap.servers";
+    private static final String TOPIC_NAMES_PROPERTY = "quarkus.kafka.streams.topics";
+    private static final String STREAMS_OPTION_PREFIX = "quarkus.kafka.streams.";
```
This should be handled by a proper config object.

For the options, you can use a `Map` and then push all the options.

I'm also wondering why the first one doesn't have a prefix? I suppose there's a reason but it's really an issue IMHO.

gunnarmorling replied:
Yes, `kafka.bootstrap.servers` has been defined by @cescoffier and also is supported by Reactive Messaging. I'm not sure whether he defined other "kafka..." options. In that light, should we perhaps simply use "kafka.streams." as a prefix? Might be better actually.

gsmet commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsTopologyManager.java:
```diff
@@ -0,0 +1,167 @@
+    private static final String TOPIC_NAMES_PROPERTY = "quarkus.kafka.streams.topics";
+    private static final String STREAMS_OPTION_PREFIX = "quarkus.kafka.streams.";
+
+    private static final String DEFAULT_BOOTSTRAP_SERVERS = "localhost:9012";
```
This could be a default value of your config item.

gsmet commented on integration-tests/kafka/src/main/java/io/quarkus/it/kafka/streams/KafkaStreamsEndpoint.java:
```diff
@@ -0,0 +1,48 @@
+    private ReadOnlyKeyValueStore<Integer, Long> getCountstore() {
+        while (true) {
+            try {
+                return streams.store("countstore", QueryableStoreTypes.keyValueStore());
```
Hmmm. I see you remove awaitility but I would really prefer we use that rather than an unbounded loop.

gunnarmorling replied:
This one didn't use Awaility before, it's just extracted from the previous common class for pipeline and endpoint.

gsmet reviewed (commented):
I added a couple of comments inline.

gunnarmorling commented:
@gsmet, would you have an example for reading all the options into a map? As per my latest mail on the list, I think those "pass-through" options should not be using the "quarkus" prefix, so I'm not sure how the ConfigGroup stuff would look like for that. Thanks for any pointers!

gunnarmorling commented:
I've pushed an update: runtime init changeable properties are given with the "quarkus.kafka-streams" prefix now, and using the "-" as separator.

All others are given using the "kafka-streams" prefix and their original name, i.e. with "." as separator:

```
quarkus.kafka-streams.bootstrap-servers=localhost:19092
quarkus.kafka-streams.application-id=streams-test-pipeline
quarkus.kafka-streams.topics=streams-test-categories,streams-test-customers

kafka-streams.cache.max.bytes.buffering=...
```

That's how I'd interpret the wording of the extension guide and what @emmanuelbernard told me about it. For reading the latter, I'm manually loading application.properties at build time and pass the loaded `Properties` to the template. No idea whether it's the right way to do this or not, but I didn't find anything else. Happy to alter if there's preferred alternative.

I'll review the existing options and likely add a few more to the runtime init changeable group.

@gsmet, @cescoffier please let me know what you think. Thanks!

machi1990 commented:
Hey @gunnarmorling 
> so I'm not sure how the ConfigGroup stuff would look like for that. 

The map config in `KafkaStreamsRuntimeConfig` will look like.  
```
@ConfigItem(name = ConfigItem.PARENT) // keep the name of the parent 
public Map<String, String> streamsOptions;
```
And the `application.properties` will look like:
```
# streams options
quarkus.kafka-streams."cache.max.bytes.buffering"=10240
quarkus.kafka-streams."commit.interval.ms"=1000
quarkus.kafka-streams."metadata.max.age.ms"=500
quarkus.kafka-streams."auto.offset.reset"=earliest
... etc
```
HTH

machi1990 commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * A
+     */
+    public Optional<String> topics;
```
`@ConfigItem`  missing?

gunnarmorling replied:
Thanks, added it.

emmanuelbernard commented:
> I've pushed an update: runtime init changeable properties are given with the "quarkus.kafka-streams" prefix now, and using the "-" as separator.
> 
> All others are given using the "kafka-streams" prefix and their original name, i.e. with "." as separator:
> 
> ```
> quarkus.kafka-streams.bootstrap-servers=localhost:19092
> quarkus.kafka-streams.application-id=streams-test-pipeline
> quarkus.kafka-streams.topics=streams-test-categories,streams-test-customers
> 
> kafka-streams.cache.max.bytes.buffering=...
> ```
> 
> That's how I'd interpret the wording of the extension guide and what @emmanuelbernard told me about it. For reading the latter, I'm manually loading application.properties at build time and pass the loaded `Properties` to the template. No idea whether it's the right way to do this or not, but I didn't find anything else. Happy to alter if there's preferred alternative.
> 
> I'll review the existing options and likely add a few more to the runtime init changeable group.
> 
> @gsmet, @cescoffier please let me know what you think. Thanks!

@dmlloyd can we offer a facility to get all non mapped properties (i.e. non `quarkus.*`) from application.properties under a given prefix (`kafka-streams.*` in this case. It feels wrong to have to read the application.properties manually in the extension and miss the other ConfigSources.

dmlloyd commented:
> @dmlloyd can we offer a facility to get all non mapped properties (i.e. non `quarkus.*`) from application.properties under a given prefix (`kafka-streams.*` in this case. It feels wrong to have to read the application.properties manually in the extension and miss the other ConfigSources.

You should be able to do this already using the MP config API.  `org.eclipse.microprofile.config.Config#getPropertyNames` is the simplest way to do it.  Note that you can't get run time properties at build time (obviously).

gunnarmorling commented:
> You should be able to do this already using the MP config API. 

@dmlloyd, would you have a pointer on getting an instance of the MP `Config` at build time? Agreed that this would be the simplest and also most complete.

dmlloyd commented:
It should be as simple as calling `ConfigProvider.getConfig()`.

gsmet commented:
I know we already had a similar discussion for Camel.

I don't have an issue in having a separate namespace for pass-through config. That being said, I'm wondering if we should have some support for it in the config code. Otherwise we will end up filtering the whole config for each of these properties.

Wondering if we could somehow have something to get the config properties of a given namespace (so `quarkus`, `camel`, `kafka-streams`).

@dmlloyd does it make sense?

dmlloyd commented:
You'd only have to filter it one time, since it's basically an iterator.  You could construct a sub-map which could be then used for the pass-through properties.  Of course it depends on how the given project consumes its properties.

gsmet commented:
When I mean several times, it's one time per extension consuming them.

Not sure if we would use Camel and Kafka Streams together but if we generalize this, we will have more extensions doing that.

dmlloyd commented:
I would hope that this is not something we do often.  But I don't think the cost is a very big deal, since you'd still be doing it a small constant number of times even if you have several extensions doing it.

gunnarmorling commented:
> It should be as simple as calling ConfigProvider.getConfig().

LOL, this was too easy. I thought I'd get it via some build item, context or something. Yes, this works. Thanks, @dmlloyd.

gsmet commented:
Sorry I didn't have the time to drive this puppy home before the release, I'm marking it for the next one so that we keep an eye on it.

BTW, it needs a rebase.

gunnarmorling commented:
No worries, I was busy with other things myself, too. I'll rebase and also need to pick up some loose ends like the JNI stuff you mentioned. Would be nice to have it for the next release.

gunnarmorling committed:
#2863 Representing bootstrap servers as List<InetSocketAddress>

gunnarmorling committed:
#2863 Simplifying native builds

gunnarmorling commented:
Hi @gsmet , this is rebased and updated now. I think it's mostly good to go; most importantly I'll need to update the quickstart guide accordingly, it will be quite a bit simpler now. But I don't think this has to stop this PR from progressing. I've implemented what was discussed on the mailing list:

* a small set of runtime-configurable variables via a config group `quarkus.kafka-streams...`, and pass-through variables for everything else under `kafka-streams...`. The former apply the Quarkus naming conventions, the latter don't. I hope that makes sense.

dmlloyd commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java:
```diff
@@ -0,0 +1,42 @@
+     * A unique identifier for this Kafka Streams application.
+     */
+    @ConfigItem
+    public String applicationId;
```
Is this something that could potentially be defaulted to e.g. `${quarkus.application.name}`?

gunnarmorling replied:
Yes, good idea; I wasn't aware of this property, but it seems like a good default. Will add a commit for that.

gunnarmorling replied:
@dmlloyd, so I tried and changed it into this:

    @ConfigItem(defaultValue = "${quarkus.application.name}")
    public String applicationId;

But this gives me an error upon start similar to the one we discussed before:

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 9.66 s <<< FAILURE! - in io.quarkus.it.kafka.KafkaConsumerTest
[ERROR] test  Time elapsed: 0.01 s  <<< ERROR!
org.junit.jupiter.api.extension.TestInstantiationException:
TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [io.quarkus.it.kafka.KafkaConsumerTest]: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.application.name not found
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestInstanceFactory(ClassBasedTestDescriptor.java:298)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:273)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:70)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:256)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$2(ClassBasedTestDescriptor.java:250)
	at java.base/java.util.Optional.orElseGet(Optional.java:369)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$3(ClassBasedTestDescriptor.java:249)
	at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:29)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:106)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:105)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:69)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
Caused by: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.application.name not found
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:241)
	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:311)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestInstanceFactory(ClassBasedTestDescriptor.java:283)
	... 47 more
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.application.name not found
	at io.quarkus.builder.Execution.run(Execution.java:108)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:121)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:118)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:101)
	... 50 more
Caused by: java.util.NoSuchElementException: Property quarkus.application.name not found
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:89)
	at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:55)
	at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:15)
	at org.wildfly.common.expression.ExpressionNode.emit(ExpressionNode.java:42)
	at org.wildfly.common.expression.Expression.evaluateException(Expression.java:75)
	at org.wildfly.common.expression.Expression.evaluate(Expression.java:89)
	at io.quarkus.runtime.configuration.ExpandingConfigSource.expandValue(ExpandingConfigSource.java:76)
	at io.quarkus.deployment.configuration.ObjectConfigType.getDefaultValueIntoEnclosingGroup(ObjectConfigType.java:35)
	at io.quarkus.deployment.configuration.GroupConfigType.create(GroupConfigType.java:142)
	at io.quarkus.deployment.configuration.GroupConfigType.getOrCreate(GroupConfigType.java:201)
	at io.quarkus.deployment.configuration.ConfigDefinition.initialize(ConfigDefinition.java:164)
	at io.quarkus.deployment.configuration.ConfigDefinition.loadConfiguration(ConfigDefinition.java:385)
	at io.quarkus.deployment.steps.ConfigurationSetup.initializeConfiguration(ConfigurationSetup.java:273)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:508)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1429)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

Note this is *with* your fix https://github.com/dmlloyd/quarkus/commit/fix-3030-again. Am I doing something generally wrong here?

dmlloyd replied:
I guess it's a result of #3008.  The `quarkus.application.name` property is done in a wrong/hacky way, which means it doesn't get into the default values set.

gunnarmorling replied:
Coming back to this; as #3008/#1829 still is open, the `quarkus.kafka-streams.application-id` will be mandatory for now. Once #1829 is resolved, we can make it optional, defaulting to `quarkus.application.name`.

gunnarmorling replied:
I've added this as one todo in #3500.

dmlloyd commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java:
```diff
@@ -0,0 +1,42 @@
+     * A comma-separated list of host:port pairs identifying the Kafka bootstrap server(s)
+     */
+    @ConfigItem(defaultValue = "localhost:9012")
+    public String bootstrapServers;
```
Would a `List<InetSocketAddress>` be more appropriate here?  That would give some base level of validation.  Or at a minimum a `List<String>`?

gunnarmorling replied:
Would a value like "host1:9012,host2:9012" automatically parsed into `List<InetSocketAddress>`? Then it seems like a good thing to do; `List<String>` doesn't seem overly useful, would not essentially every value qualify as such? Note the value must be passed as a string to Kafka Streams, so the conversion will be a bit of a forth and back, but if `List<InetSocketAddress>` works, why not.

dmlloyd replied:
> Would a value like "host1:9012,host2:9012" automatically parsed into `List<InetSocketAddress>`?

Yes, exactly.

> Note the value must be passed as a string to Kafka Streams, so the conversion will be a bit of a forth and back, but if `List<InetSocketAddress>` works, why not.

It'll be some free fail-fast validation.  At some point we'll have a consolidated config validation blob that will print out nice errors like "xxxx is not a valid value for properties of type `InetSocketAddress`" or something like that.  You'll be ready for it!

gunnarmorling replied:
Done.

dmlloyd commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java:
```diff
@@ -0,0 +1,42 @@
+     * A unique identifier of this application instance, typically in the form host:port.
+     */
+    @ConfigItem
+    public Optional<String> applicationServer;
```
How does this compare to `applicationId` above?

gunnarmorling replied:
`applicationId` is used to identify different consumers of a Kafka topic. This lets multiple applications (or, more generically, consumer groups in Kafka lingo) consume one topic, ensuring that each message is consumed by each consumer group. So if you had two different Kafka Streams applications processing the same topic, they'd have to use different application ids.

`applicationServer` is used to identify individual nodes of one clustered Kafka Streams application. In this case each instance processes a subset of the partitions of the source topic(s). Kafka Streams provides the metadata APIs for finding out which instance of the clustered app is in charge of which messages (by means of assigning message keys to partitions), so that a consumer can identify the right node in a cluster for getting the state for a given key.

dmlloyd replied:
Ah okay.  I didn't really get that from the property names or descriptions.

gunnarmorling replied:
Gotcha; to be fair I think it'll be apparent to Kafka Streams users. I'll also make the mapping apparent in the docs.

gunnarmorling committed:
#2863 Documentation update

gunnarmorling commented:
@gsmet Rebased and updated. I've extracted issue #3500 from #2863 for tracking all those things that are not addressed yet, but also the changes done present a great simplification of using this. I'll file a PR for updating the quickstart repo shortly.

gunnarmorling committed:
#2863 Adding missing @RegisterForReflection

gsmet commented on docs/src/main/asciidoc/kafka-streams-guide.adoc:
```diff
@@ -412,69 +366,57 @@ public class KafkaStreamsPipeline {
+<2> The `temperature-values` topic is read into a `KStream`; whenever a new message arrives to this topic, the pipeline will be processed for this measurement
+<3> The message from the `temperature-values` topic is joined with the corresponding weather station, using the topic's key (weather station id); the join result contains the data from the measurement and associated weather station message
+<4> The values are grouped by message key (the weather station id)
+<5> Within each group, all the measurements of that station are aggregated, by keeping track of minimum and maxium values and calculating the average value of all measurements of that station (see the `Aggregation` type)
```
```suggestion
<5> Within each group, all the measurements of that station are aggregated, by keeping track of minimum and maximum values and calculating the average value of all measurements of that station (see the `Aggregation` type)
```

gsmet commented on docs/src/main/asciidoc/kafka-streams-guide.adoc:
```diff
@@ -412,69 +366,57 @@ public class KafkaStreamsPipeline {
-                        topicNames.contains(TEMPERATURE_VALUES_TOPIC)) {
-                    return;
-                }
+The options with the `quarkus.kafka-streams` prefix can be changed dynamically at application start-up,
```
```suggestion
The options with the `quarkus.kafka-streams` prefix can be changed dynamically at application startup,
```
?

gsmet commented on docs/src/main/asciidoc/kafka-streams-guide.adoc:
```diff
@@ -412,69 +366,57 @@ public class KafkaStreamsPipeline {
+e.g. via environment variables or system properties.
+`bootstrap-servers`, `application-id` and `application-server` are mapped to the Kafka Streams properties `bootstrap.servers`, `application.id` and `application.server`, respectively.
+`topics` is specific to Quarkus: the application will wait for all the given topics to exist before launching the Kafka Streams engine.
+This is to done to gracefully awati the creation of topics that don't yet exist at application start-up time.
```
```suggestion
This is to done to gracefully await the creation of topics that don't yet exist at application startup time.
```

gsmet commented on extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java:
```diff
@@ -0,0 +1,44 @@
+
+    /**
+     * A comma-separated list of topic names processed by this stream processing application.
+     * The pipeline will only be started once all thes topics are present in the Kafka cluster.
```
```suggestion
     * The pipeline will only be started once all these topics are present in the Kafka cluster.
```

gsmet reviewed (changes requested):
Looks good to me.

I spotted a few typos, could you fix them?

gunnarmorling committed:
Apply suggestions from code review

#2863 Typo fixes

Co-Authored-By: Guillaume Smet <guillaume.smet@gmail.com>
